src/components/organisms/SchedulePageModal/SchedulePageModal.utils.ts

```diff
@@ -0,0 +1,97 @@
+import {
+  addDays,
+  compareAsc,
+  differenceInMinutes,
+  endOfDay,
+  format,
+  isWithinInterval,
+  startOfDay,
+} from "date-fns";
+
+import { Room } from "types/rooms";
+import { PersonalizedVenueEvent, VenueEvent } from "types/venues";
+import { MyPersonalizedSchedule } from "types/User";
+
+import { WithVenueId } from "utils/id";
+import { ONE_MINUTE_IN_SECONDS, ONE_SECOND_IN_MILLISECONDS } from "utils/time";
+import { isTruthy } from "utils/types";
+
+import { RoomWithEvents } from "components/molecules/Schedule/Schedule.types";
+
+export const isEventThisDay = (date: Date) => {
+  return (event: VenueEvent) => {
+    return isWithinInterval(date, {
+      start: startOfDay(event.start_utc_seconds * ONE_SECOND_IN_MILLISECONDS),
+      end: endOfDay(
+        (event.start_utc_seconds +
+          event.duration_minutes * ONE_MINUTE_IN_SECONDS) *
+          ONE_SECOND_IN_MILLISECONDS
+      ),
+    });
+  };
+};
```

Comment:
This seems very similar to a function I saw in `utils/time` I believe? (and I noted some ways to simplify/improve it there as well)

Also, any reason we have this here rather than in `utils/time`?

Reply:
If you remove the extra `{` `}` then you can also remove the extra `return` statements and it will make things a lot cleaner.

eg.

```typescript
export const isEventThisDay = (date: Date) => (event: VenueEvent) => isWithinInterval(
```

Reply:
Also.. is the logic here correct? The name implies that we want to be checking if the event is within the specified day, yet the logic seems to be checking if the provided date is within the event duration.

Reply:
changed the logic a bit. Please, check `isEventLaterThisDay` : https://github.com/sparkletown/sparkle/pull/1302/commits/9fce282d69bf3236a3cad6381d0b990072e0c3c2

Reply:
> The name implies that we want to be checking if the event is within the specified day, yet the logic seems to be checking if the provided date is within the event duration.

I'm still not sure that the logic is correct here based on what the function name implies? Alternatively, the function name is wrong?

Can you describe what the intended logic of this function is?

Reply:
variable `date` in the arguments is either `now` (if 'Today' is selected) or `startOfDay()`.

I tried to visualize the function (to be more precise, today on the picture == selected day): 
![image](https://user-images.githubusercontent.com/63313732/117409715-7d6a7180-aedf-11eb-9415-01de4c87bf51.png)

We need to filter out all the events that are started and finished earlier than today(1) and those which are started next day and later (4), and leave cases (2) and (3).

So we artificially start all the events at the beginning of the day and then we check if selected `startOfDay()` is in the interval (startOfEventDay --- endOfEvent).

This way we get all the events that are going on the needed / selected day.

Reply:
@mike-lvov Can you please take a look at this and see if you can follow the logic/it makes sense to you/if you have any better suggestions. I still can't wrap my head around it, it feels like it could be made better, but because I can't mentally process it, I can't give any better suggestions.

Reply:
@sunny-viktoryia the name of the function says that you are checking if the event is within a specific date, but the logic is wether the day is within the event length, which seems wrong?

I guess you should be checking at least two scenarios:
 If the start of the event is within the day's limits 
 or the end of the event is within the day's limits.
 
Plus the naming of the function says it's a boolean, not a function. 
`CheckIfEventHappensWithinDay` is a better choice @sunny-viktoryia, @0xdevalias ???

And is there a reason to have it as `(day) => (event) => boolean` rather than `(day, event) => boolean`?

Reply:
> And is there a reason to have it as (day) => (event) => boolean rather than (day, event) => boolean?

curried function is used here because I need a function in the filter. Here is the usage:
```js
events.filter(
  isEventLaterThisDay(selectedDayIndex === 0 ? Date.now() : dayStart)
)
```

> I guess you should be checking at least two scenarios:
If the start of the event is within the day's limits
or the end of the event is within the day's limits.

is that enough? Should we cover the cases when the event is really long and can last several days? If we check only startDate and endDate we won't be able see it during the days between startDate and endDate.

>  CheckIfEventHappensWithinDay is a better choice @sunny-viktoryia, @0xdevalias ???
> I usually call these curried functions ‘makers’, so it would end up something like ‘makeIsEventLaterThisDay’ or similar.

I can do something like: `makeCheckIfEventHappensLaterThisDay` (is that okay?)

`CheckIfEventHappensWithinDay` - is not 100% correct, because if we specify `Date.now()` as a parameter (as I do for today's events), it'll show only the events that are going now or later today.

Reply:
> > I guess you should be checking at least two scenarios:
> > If the start of the event is within the day's limits
> > or the end of the event is within the day's limits.
> 
> is that enough? Should we cover the cases when the event is really long and can last several days? If we check only startDate and endDate we won't be able see it during the days between startDate and endDate.

Seems like a valid case that we would want to handle.

@mike-lvov did say '**at least** two scenarios'. So there is probably also the edgecase you called out above @sunny-viktoryia

---

RE: https://github.com/sparkletown/sparkle/pull/1302#discussion_r628003146

> We need to filter out all the events that are started and finished earlier than today(1) and those which are started next day and later (4), and leave cases (2) and (3).

@sunny-viktoryia So if I was to try and state the intent of this function in a single high level sentence, it's to check whether some part of the event occurs between the start and end of the provided day?

If so, I think `isEventWithinDay(date: number | Date)` would make sense to me from a naming perspective.

Then within that I would be doing something like:

- getting `startOfDay(date)`, `endOfDay(date)`, and then make it an interval
- make an interval from the start/end of the event
- check if the intervals overlap
  - I wouldn't use this lib directly, but here's some seemingly basic logic for checking if intervals overlap.
    - https://github.com/saintplay/are-intervals-overlapping/blob/master/src/index.js#L61-L76
  - I would create a generic function in `utils/time` that checks if intervals overlap based on this
    - For naming, could call it something like `checkIfIntervalsIntersect` or `doIntervalsIntersect` or similar (as it's basically an intersection: https://en.wikipedia.org/wiki/Intersection_(set_theory))

Reply:
In fact, even better, it looks like `date-fns` already handles this for us:

- https://date-fns.org/v2.21.3/docs/areIntervalsOverlapping
  - > Is the given time interval overlapping with another time interval? Adjacent intervals do not count as overlapping.

So using the above + the following:

- https://date-fns.org/v2.21.3/docs/Interval
- https://date-fns.org/v2.21.3/docs/startOfDay
- https://date-fns.org/v2.21.3/docs/endOfDay

I would expect something like this should probably cover it + all the edgecases?

```typescript
export const isEventWithinDate = (checkDate: number | Date) => (
  event: VenueEvent
) => {
  const checkDateInterval = {
    start: startOfDay(checkDate),
    end: endOfDay(checkDate),
  }

  const eventInterval = todoThatHelperFunctionThatCalculatesThis(event);

  return areIntervalsOverlapping(checkDateInterval, eventInterval)
}
```

Reply:
indeed, mastering `date-fns` can work a miracle. :upside_down_face: 
thanks!

refactored here: https://github.com/sparkletown/sparkle/pull/1302/commits/799902a40ba07362f18780b48b286cb984dadd28